### PR TITLE
refactor: add typescript types to TreeNode and TreeView 

### LINKED ISSUE #13581: Add TypeScript types to TreeView, TreeNode

### PR DESCRIPTION
**Pull request by cesardlinx**
Closes #13581

Typescript annotations added to TreeNode and TreeView 

#### Changelog

**Changed**

- TreeNode and TreeView components now use Typescript annotations
- Renamed TreeNode and TreeView file extension to .tsx
- Renamed index file extension to .ts

**Removed**

- n/a

#### Testing / Reviewing

- I run the tests for TreeNode and TreeView and it passes, also I checked that the components are still working in Storybook

image

### CODE REVIEW

**Comment by cesardlinx** [comment]
I have read the DCO document and I hereby sign the DCO.

File: packages/react/src/components/TreeView/TreeNode.tsx [hunk 1]
```diff
@@ -5,14 +5,168 @@
...
+interface TreeNodeProps {
+  /**
+   * Node ID
+   */
+  id: string | number;
+
+  /**
+   * The value of the active node in the tree
+   */
+  active: string | number;
```
Comment (tw15egan) [inline packages/react/src/components/TreeView/TreeNode.tsx]: Can we keep the props `isRequired` values the same as the PropType definitions? Seems like this is enforcing a lot of optional props as required.

Comment (cesardlinx) [reply packages/react/src/components/TreeView/TreeNode.tsx]: Absolutely @tw15egan , I've already made the necessary changes to align the `isRequired` values with the PropType definitions. This should no longer enforce optional props as required. Thanks for catching that!

**Comment by andreancardona** [comment]
@cesardlinx Hey David, do you mind fixing the merge conflicts - we'll re-review and get it merged!

**Comment by cesardlinx** [comment]
Hey @andreancardona sure thing! I'll fix it today 🙂

**Comment by cesardlinx** [comment]
Hey @andreancardona I have fixed the merge conflict, please let me know if you need anything else

File: packages/react/src/components/TreeView/index.ts [hunk 2]
```diff
@@ -8,7 +8,5 @@
 import TreeView from './TreeView';
 import TreeNode from './TreeNode';
 
-TreeView.TreeNode = TreeNode;
-
```
Comment (tay1orjones) [inline packages/react/src/components/TreeView/index.ts]: I think we'll need to keep this in to avoid causing a breaking change for anyone using `<TreeView.TreeNode>`

File: packages/react/src/components/TreeView/TreeView.tsx [hunk 3]
```diff
@@ -104,6 +179,7 @@ export default function TreeView({
         keys.ArrowDown,
         keys.Home,
         keys.End,
+        // @ts-ignore
```
Comment (tay1orjones) [inline packages/react/src/components/TreeView/TreeView.tsx:182]: What error is being suppressed here?

File: packages/react/src/components/TreeView/TreeView.tsx [hunk 4]
```diff
@@ -119,8 +195,9 @@ export default function TreeView({
     if (match(event, keys.ArrowDown)) {
       nextFocusNode = treeWalker.current.nextNode();
     }
+    // @ts-ignore
```
Comment (tay1orjones) [inline packages/react/src/components/TreeView/TreeView.tsx:198]: What error is being suppressed here?

File: packages/react/src/components/TreeView/TreeNode.tsx [hunk 5]
```diff
@@ -203,13 +373,19 @@ const TreeNode = React.forwardRef(
...
+        aria-current={treeNodeProps['aria-current'] || undefined}
+        aria-selected={treeNodeProps['aria-selected'] || undefined}
+        id={treeNodeProps.id as string}>
         <div className={`${prefix}--tree-node__label`} ref={currentNodeLabel}>
           {/* https://github.com/carbon-design-system/carbon/pull/6008#issuecomment-675738670 */}
           {/* eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions */}
           <span
             className={`${prefix}--tree-parent-node__toggle`}
-            disabled={disabled}
+            // disabled={disabled}
```
Comment (tay1orjones) [inline packages/react/src/components/TreeView/TreeNode.tsx]: I think this needs to remain

File: packages/react/__tests__/__snapshots__/PublicAPI-test.js.snap [hunk 6]
```diff
@@ -9060,89 +9057,6 @@ Map {
     "render": [Function],
   },
   "TreeView" => Object {
-    "TreeNode": Object {
-      "$$typeof": Symbol(react.forward_ref),
-      "propTypes": Object {
-        "active": Object {
```
Comment (tay1orjones) [inline packages/react/__tests__/__snapshots__/PublicAPI-test.js.snap]: Once `TreeView.TreeNode` is added back in, regenerate snapshots with `yarn test -u` and this change should be reverted/go away.

File: packages/react/src/components/TreeView/TreeNode.tsx [hunk 7]
```diff
@@ -5,15 +5,163 @@
...
+   */
+  'aria-selected'?: boolean | null;
+
+  /**
+   * Indicates that the element is perceivable but disabled
+   */
+  'aria-disabled'?: boolean;
+
+  /**
+   * Name of a role in the ARIA specificatio
```
Comment (tay1orjones) [inline packages/react/src/components/TreeView/TreeNode.tsx]:
```suggestion
   * Name of a role in the ARIA specification
```

File: packages/react/src/components/TreeView/TreeNode.tsx [hunk 8]
```diff
@@ -151,20 +314,22 @@ const TreeNode = React.forwardRef(
...
-          return depth + 1 + depth * 0.5;
-        }
-        // parent node without icon
-        if (children) {
-          return depth + 1;
-        }
-        // leaf node with icon
-        if (Icon) {
-          return depth + 2 + depth * 0.5;
+        if (depth != null) {
```
Comment (tay1orjones) [inline packages/react/src/components/TreeView/TreeNode.tsx:317]: I don't think this is necessary since you updated the default function parameter to always set it to `0`?

File: packages/react/src/components/TreeView/TreeNode.tsx [hunk 9]
```diff
@@ -203,13 +373,19 @@ const TreeNode = React.forwardRef(
...
     return (
       // eslint-disable-next-line jsx-a11y/role-supports-aria-props
-      <li {...treeNodeProps} aria-expanded={!!expanded} ref={ref}>
+      <li
+        {...treeNodeProps}
+        aria-expanded={!!expanded}
+        ref={ref}
+        aria-current={treeNodeProps['aria-current'] || undefined}
+        aria-selected={treeNodeProps['aria-selected'] || undefined}
+        id={treeNodeProps.id as string}>
```
Comment (tay1orjones) [inline packages/react/src/components/TreeView/TreeNode.tsx:383]: Why is this necessary instead of just leaving these within `{...treeNodeProps}`?

File: packages/react/src/components/TreeView/TreeNode.tsx [hunk 10]
```diff
@@ -288,7 +464,7 @@ TreeNode.propTypes = {
    * Optional prop to allow each node to have an associated icon.
    * Can be a React component class
    */
-  renderIcon: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
```
Comment (tay1orjones) [inline packages/react/src/components/TreeView/TreeNode.tsx]: Narrowing proptypes is considered a breaking change. We'll need to leave this as is for now.

File: packages/react/src/components/TreeView/TreeView.tsx [hunk 11]
```diff
@@ -148,6 +225,7 @@ export default function TreeView({
           }
         }
       }
+      // @ts-ignore
```
Comment (tay1orjones) [inline packages/react/src/components/TreeView/TreeView.tsx:228]: What error is being suppressed here?

**Review by tay1orjones — CHANGES_REQUESTED** [review]
Thanks for diving in to this! I have just a few observations below.

File: packages/react/src/components/TreeView/TreeView.tsx (hunk 3 above)
Comment (cesardlinx) [reply packages/react/src/components/TreeView/TreeView.tsx:182]: This is the error I'm suppressing here.
image

File: packages/react/src/components/TreeView/TreeView.tsx (hunk 4 above)
Comment (cesardlinx) [reply packages/react/src/components/TreeView/TreeView.tsx:198]: The error here is the same as in the previous comment
image

File: packages/react/src/components/TreeView/TreeNode.tsx (hunk 5 above)
Comment (cesardlinx) [reply packages/react/src/components/TreeView/TreeNode.tsx]: If I add it back I get this error
image

Basically what it is trying to say is that span element doesn't have a `disabled` prop. And it makes sense because only html elements such as button, input, select, textarea have the `disabled` attribute

File: packages/react/src/components/TreeView/TreeNode.tsx (hunk 8 above)
Comment (cesardlinx) [reply packages/react/src/components/TreeView/TreeNode.tsx:317]: Here I'm performing the extra check because typescript detects it as possibly 'undefined'
image

File: packages/react/src/components/TreeView/TreeNode.tsx (hunk 9 above)
Comment (cesardlinx) [reply packages/react/src/components/TreeView/TreeNode.tsx:383]: I'm glad you asked, this is necessary in order to make the types match in this section, otherwise we will get this error:
image

File: packages/react/src/components/TreeView/TreeNode.tsx (hunk 10 above)
Comment (cesardlinx) [reply packages/react/src/components/TreeView/TreeNode.tsx]: The problem was that by leaving it as how it was I get this error:
image

It appears to be a conflict between TypeScript and PropTypes in this section. The closest issue I found resembling this problem is documented here: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/28015

File: packages/react/src/components/TreeView/TreeView.tsx (hunk 11 above)
Comment (cesardlinx) [reply packages/react/src/components/TreeView/TreeView.tsx:228]: Here I'm suppressing the same error as before
image

**Comment by cesardlinx** [comment]
Thank you @tay1orjones for the valuable feedback. I've addressed some of the issues you highlighted, and also I've provided responses to your observations.

File: packages/react/src/components/TreeView/TreeNode.tsx (hunk 5 above)
Comment (tay1orjones) [reply packages/react/src/components/TreeView/TreeNode.tsx]: Oh my bad, yeah let's just remove it then

File: packages/react/src/components/TreeView/TreeNode.tsx (hunk 8 above)
Comment (tay1orjones) [reply packages/react/src/components/TreeView/TreeNode.tsx:317]: Ah weird, yeah I see on second glance it's actually not the default function parameter that you updated, it's just the `depth: (depth ?? 0)` in the `cloneElement` call. This looks good to me then 👍

File: packages/react/src/components/TreeView/TreeView.tsx (hunk 3 above)
Comment (wkeese) [reply packages/react/src/components/TreeView/TreeView.tsx:182]: It's too bad that match.js hasn't been converted to Typescript yet, but I think you can remove the @ts-ignore by just fixing match.js:

```js
/**
 * @typedef Key
 * @property key {Array<string>|string}
 * @property which {number}
 * @property keyCode {number}
 *  @property code {string}        <-- new line here
 */
```

**Comment by tay1orjones** [comment]
Hey there, sorry for the delay in response here. From what I can tell there's two pieces we need to resolve here

1. Remove the `disabled` from TreeNode instead of commenting it out
2. Find a workaround for `renderIcon` in TreeNode because we can't narrow the Proptypes without a breaking change

**Comment by cesardlinx** [comment]
Hey @tay1orjones. Sorry, work has kept me busy lately and I almost forgot about this PR. I'll check it out during the weekend and I'll keep you posted.

**Comment by cesardlinx** [comment]
Hey @tay1orjones, I've resolved the issues you pointed out. Let me know if there's anything else I can help with.

**Comment by Gururajj77** [comment]
Hi @cesardlinx,

Apologies for the delay. Could you please resolve the merge conflicts so that we can proceed?

Thank you.

**Comment by tay1orjones** [comment]
This was put together before the addition of the TreeView controllable api, which after resolving the merge conflicts there's many tests failing. We're going to close this one in favor of opening a new one based on these changes but with the controllable api taken into consideration. Sorry for the delay here
